sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp

```diff
@@ -0,0 +1,1006 @@
+      if constexpr (detail::is_sub_group<GroupTy>::value) {
+        if (group.get_local_range() != group.get_max_local_range())
+          // Sub-group is not "full".
+          return generic();
+      } else if constexpr (detail::is_group<GroupTy>::value) {
+        // TODO: Use get_child_group from
+        // sycl_ext_oneapi_root_group extension once it is
+        // implemented instead of this free function.
+        auto ndi = sycl::ext::oneapi::experimental::this_nd_item<
+            GroupTy::dimensions>();
+        auto sg = ndi.get_sub_group();
+
+        auto wg_size = group.get_local_range().size();
+        auto simd_width = sg.get_max_local_range().size();
+
+        if (wg_size % simd_width != 0) {
+          return generic();
+        }
+      }
+    }
+
+    // TODO: verify that sizeof * sg.get_max_local_range() *
+    // sg_offset_per_wi is a multiple of required alignment for
+    // detail::is_group.
+
+    constexpr auto AS = detail::deduce_AS<iter_no_cv>::value;
+    if constexpr (AS == access::address_space::global_space) {
+      // The only customization point - to be handled by the
+      // caller.
+      return impl(iter);
+    } else if constexpr (AS == access::address_space::generic_space) {
+      if (auto global_ptr =
+              __SYCL_GenericCastToPtrExplicit_ToGlobal<value_type>(iter))
+        return delegate(global_ptr);
+      else
+        return generic();
+    } else {
+      return generic();
+    }
+  }
+  __builtin_unreachable();
+}
+#endif
+} // namespace detail
+
+#ifdef __SYCL_DEVICE_ONLY__
+// Load API scalar.
+template <typename Group, typename InputIteratorT, typename OutputT,
+          typename Properties = decltype(properties())>
+std::enable_if_t<
+    std::is_convertible_v<remove_decoration_t<typename std::iterator_traits<
+                              InputIteratorT>::value_type>,
+                          OutputT> &&
+    detail::is_generic_group_v<Group>>
+group_load(Group g, InputIteratorT in_ptr, OutputT &out,
+           Properties properties = {}) {
+  group_load(g, in_ptr, span<OutputT, 1>(&out, 1), properties);
+}
+
+// Store API scalar.
+template <typename Group, typename InputT, typename OutputIteratorT,
+          typename Properties = decltype(properties())>
+std::enable_if_t<std::is_convertible_v<
+                     InputT, remove_decoration_t<typename std::iterator_traits<
+                                 OutputIteratorT>::value_type>> &&
+                 detail::is_generic_group_v<Group>>
+group_store(Group g, const InputT &in, OutputIteratorT out_ptr,
+            Properties properties = {}) {
+  group_store(g, span<const InputT, 1>(&in, 1), out_ptr, properties);
+}
+
+// Load API sycl::vec overload.
+template <typename Group, typename InputIteratorT, typename OutputT, int N,
+          typename Properties = decltype(properties())>
+std::enable_if_t<
+    std::is_convertible_v<remove_decoration_t<typename std::iterator_traits<
```

Comment:
Will remove_decoration_t remove cv-qualifiers? I noticed compilation issues when sycl::buffer with read_only access is used as an input